src/core/qgsofflineediting.cpp

```diff
@@ -359,16 +266,33 @@ void QgsOfflineEditing::synchronize()
         // Invalidate the connection to force a reload if the project is put offline
         // again with the same path
         offlineLayer->dataProvider()->invalidateConnections( QgsDataSourceUri( offlineLayer->source() ).database() );
-        // remove offline layer
-        QgsProject::instance()->removeMapLayers( QStringList() << qgisLayerId );
 
-
-        // disable offline project
-        QString projectTitle = QgsProject::instance()->title();
-        projectTitle.remove( QRegularExpression( " \\(offline\\)$" ) );
-        QgsProject::instance()->setTitle( projectTitle );
-        QgsProject::instance()->removeEntry( PROJECT_ENTRY_SCOPE_OFFLINE, PROJECT_ENTRY_KEY_OFFLINE_DB_PATH );
         remoteLayer->reload(); //update with other changes
+        offlineLayer->setDataSource( remoteLayer->source(), remoteLayer->name(), remoteLayer->dataProvider()->name() );
+
+
+        // remove offline layer properties
+        layer->removeCustomProperty( CUSTOM_PROPERTY_IS_OFFLINE_EDITABLE );
+
+        // remove original layer source and information
+        layer->removeCustomProperty( CUSTOM_PROPERTY_REMOTE_SOURCE );
+        layer->removeCustomProperty( CUSTOM_PROPERTY_REMOTE_PROVIDER );
+        layer->removeCustomProperty( CUSTOM_PROPERTY_ORIGINAL_LAYERID );
+        layer->removeCustomProperty( CUSTOM_PROPERTY_LAYERNAME_SUFFIX );
+
+        // remove connected signals
+        disconnect( layer, &QgsVectorLayer::editingStarted, this, &QgsOfflineEditing::startListenFeatureChanges );
+        disconnect( layer, &QgsVectorLayer::editingStopped, this, &QgsOfflineEditing::stopListenFeatureChanges );
+
+        //add constrainst of fields that use defaultValueClauses from provider on original
+        const auto fields = remoteLayer->fields();
+        for ( const QgsField &field : fields )
+        {
+          if ( !remoteLayer->dataProvider()->defaultValueClause( remoteLayer->fields().fieldOriginIndex( remoteLayer->fields().indexOf( field.name() ) ) ).isEmpty() )
```

Comment:
Is that the right condition here?
I.e. if there is a `now()` default clause on a dataprovider, should there be a not null constraint on the offline project?

Reply:
I want to avoid any behavior / logic change in this PR so it can be backported to 3.20 and maintain expectations by pre-existing users.

Happy to review logic choices in follow up discussions and PR :)